### PR TITLE
resolve #110: Add extra requirements on the Password CRD

### LINKED ISSUE #110: Add extra requirements on the Password CRD.
**Describe the problem/challenge you have**
The `secretgen.k14s.io/v1alpha1/Password` allows to specify the length of the generated password only. Unfortunately, external resources may ask to have other requirements such as: `Your password must contain characters from three of the following categories – English uppercase letters, English lowercase letters, numbers (0-9), and non-alphanumeric characters (!, $, #, %, etc.).` or `Your password must be at least 8 characters and at most 128 characters`

**Describe the solution you'd like**
Add extra properties to manage these requirements. The password crd can look like
```yaml
apiVersion: secretgen.k14s.io/v1alpha2
kind: Password
metadata:
  name: my-constrained-long-user-password
spec:
  length:
      min: 8
      max: 128
  categories:
     - name: uppercase        #pre-defined category
        min: 2
     - name: lowercase        #pre-defined category + min:1 default values                                              
     - name: my-custom-set
        min: 3
        available: ['!', '$', '#', '%']
```
---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you would like to work on this issue.

### PR DESCRIPTION
**Pull request by bmoussaud**
Implement issue #110 to be able to provide more inputs on the specification of the generated password.

### CODE REVIEW

File: pkg/generator/password_reconciler.go
Comment (cppforlife) [inline pkg/generator/password_reconciler.go]: math/rand unfortunately is "unsuitable for security-sensitive work" per math/rand's package documentation. even though we are using crypto/rand to select individual elements, i think shuffling needs to be "secure" as well.

**Comment by bmoussaud** [comment]
@cppforlife can you merge it?

**Comment by cppforlife** [comment]
i ve recently nudged folks to give a second set of eyes; seems like folks havent gotten any additional feedback. let me take a second look (first look i did, it looked fine to me) and we'll get it merged.

File: pkg/apis/secretgen/v1alpha1/types_password.go
Comment (cppforlife) [inline pkg/apis/secretgen/v1alpha1/types_password.go:71]: nit: it seems that it's more common to not have `charset` be "one word", meaning we should rename this key to `symbolCharset`. wdyt?

File: pkg/apis/secretgen/v1alpha1/types_password.go
Comment (cppforlife) [inline pkg/apis/secretgen/v1alpha1/types_password.go:50]: i dont think we use kubebuilder in the project, so to avoid confusion that we do i would recommend removing these comments. if we want to indicate defaults, perhaps do it in a description (comment without qualifier)?

File: pkg/generator/password_reconciler.go
Comment (cppforlife) [inline pkg/generator/password_reconciler.go]: nit: inline localGeneratePassword into  PasswordReconciler.generate -- it's not clear to me why it's needed to be separated.

File: pkg/generator/password_reconciler.go
Comment (cppforlife) [inline pkg/generator/password_reconciler.go]: seems like we we are ignoring errors when generating elements, i think they should be propagated since failure to generate char shouldnt be ignored.

File: pkg/generator/password_reconciler.go
Comment (cppforlife) [inline pkg/generator/password_reconciler.go]: can be const. also allCharSet can be next to these as well.

Comment (cppforlife) [reply pkg/generator/password_reconciler.go]: i would also recommend moving const declaration into generate function. e.g.

```go
func (r *PasswordReconciler) generate(password *sgv1alpha1.Password) (string, error) {
   const (
      lowerCharset = "..."
   )
   ...other code...
}
```

this way it's localized to where content is actually needed.

**Review by joe-kimmel-vmw — APPROVED** [review]
LGTM

File: pkg/generator/password_reconciler.go
Comment (neil-hickey) [inline pkg/generator/password_reconciler.go:159]: Is this missing the symbolcharset?

File: pkg/generator/password_reconciler.go
Comment (neil-hickey) [inline pkg/generator/password_reconciler.go:224]: Question for other maintainers, i.e @joe-kimmel-vmw / @cppforlife seems like we have a mix of upper and lower case error formatting going on. I think go conventionally pushes for the lower case, but I am see a LOT of upper cased error messaging in this repo. Any thoughts on  why we went with the uppercasing generally?

File: pkg/generator/password_reconciler.go
Comment (neil-hickey) [inline pkg/generator/password_reconciler.go:165]: Is this missing the symbolcharset? 

```suggestion
	var allCharSet = lowerCharSet + upperCharSet + digitSet + spec.SymbolCharSet
```

File: test/e2e/password_test.go
Comment (neil-hickey) [inline test/e2e/password_test.go]:
```suggestion
	name := "test-length-password-template"
```

File: test/e2e/password_test.go
Comment (neil-hickey) [inline test/e2e/password_test.go]:
```suggestion
			t.Fatalf("Expect password length is 24, found %d", len(secret.Data["val"]))
```

File: test/e2e/password_test.go
Comment (neil-hickey) [inline test/e2e/password_test.go]:
```suggestion
			t.Fatalf("Expect valx2 password length is 48, found %d", len(secret.Data["valx2"]))
```

File: test/e2e/password_test.go
Comment (neil-hickey) [inline test/e2e/password_test.go]:
```suggestion
			t.Fatalf("Expect password length is 27, found %d (value %s) ", len(secret.Data["val"]), secret.Data["val"])
```

File: test/e2e/password_test.go
Comment (neil-hickey) [inline test/e2e/password_test.go]:
```suggestion
			t.Fatalf("Expect password length is 3, found %d", len(secret.Data["val"]))
```

File: test/e2e/password_test.go
Comment (neil-hickey) [inline test/e2e/password_test.go]: We use the `"github.com/stretchr/testify/assert"` package to test generally in SGC. (I appreciate that the code has a mixture, but are moving as a project to add new code with this package used instead of t.Fatal etc)

This would actually simply alot of this testing to something like:

```suggestion
		assert.Len(t, secret.Data["val"], 3, "Expect password length is 3, found %d", len(secret.Data["val"])")
```

File: pkg/generator/password_reconciler.go
Comment (bmoussaud) [reply pkg/generator/password_reconciler.go:165]: hey, @neil-hickey: I didn't add spec.SymbolCharSet because I don't want to add extra SymbolCharset to fill in the generated password. Systems don't allow too many symbols. So if the user specifies x symbols, the user gets x symbols. not more.

Comment (neil-hickey) [reply pkg/generator/password_reconciler.go:165]: ok cool, seems reasonable to me. Just wanted to double check the reasoning 👍

File: test/e2e/password_test.go
Comment (neil-hickey) [inline test/e2e/password_test.go]:
```suggestion
		assert.Len(t, secret.Data["val"], 3, "Expect password length is 3, found %d", len(secret.Data["val"]))
```

I think my suggestion was bad, left in an extra `"`

File: pkg/generator/password_reconciler.go
Comment (cppforlife) [reply pkg/generator/password_reconciler.go:224]: carvel projects always used capital letter for starting errors.

Comment (cppforlife) [reply pkg/generator/password_reconciler.go:224]: > Any thoughts on why we went with the uppercasing generally?

i found it easier to read when k14s was started and also it was easier to split errors to be more human friendly.

**Review by cppforlife — COMMENTED** [review]
(seems like i had pending comments on this PR, probably waiting for others to reply first. submitted now.)

File: pkg/generator/password_reconciler.go
Comment (joe-kimmel-vmw) [reply pkg/generator/password_reconciler.go:224]: +1 - it's initial caps because that's what DK prefers.  I think i tried to put that in our mostly-useless style guide at some point.

File: pkg/generator/password_reconciler.go
Comment (bmoussaud) [reply pkg/generator/password_reconciler.go]: @cppforlife is it mandatory to apply this pattern before merging the pull request?

**Review by cppforlife — APPROVED** [review]
note to self: post-merge add err check for rand.Int calls

**Pull request merged**
